rust/kernel/hrtimer.rs

```diff
@@ -0,0 +1,279 @@
+//!        #[pin]
+//!        timer: Timer<Self>,
+//!        flag: Arc<AtomicBool>,
+//!    }
+//!
+//!    impl IntrusiveTimer {
+//!        fn new() -> impl PinInit<Self> {
+//!            pin_init!(Self {
+//!                timer <- Timer::new(),
+//!                flag: Arc::try_new(AtomicBool::new(false)).unwrap(),
+//!            })
+//!        }
+//!    }
+//!
+//!    impl TimerCallback for IntrusiveTimer {
+//!        type Receiver<'a> = Pin<&'a IntrusiveTimer>;
+//!
+//!        fn run(this: Self::Receiver<'_>) {
+//!            pr_info!("Timer called\n");
+//!            this.flag.store(true, Ordering::Relaxed);
+//!        }
+//!    }
+//!
+//!    impl_has_timer! {
+//!        impl HasTimer<Self> for IntrusiveTimer { self.timer }
+//!    }
+//!
+//!    stack_pin_init!(let timer = IntrusiveTimer::new());
+//!    let flag = timer.flag.clone();
+//!
+//!    timer.into_ref().schedule(200_000_000);
+//!    while !flag.load(Ordering::Relaxed) {}
+//!    pr_info!("Done\n");
+//! ```
+//!
+//! C header: [`include/linux/hrtimer.h`](srctree/include/linux/workqueue.h)
+
+use core::{marker::PhantomData, pin::Pin};
+
+use crate::{init::PinInit, prelude::*, types::Opaque};
+
+/// A timer backed by a C `struct hrtimer`
+///
+/// # Invariants
+///
+/// * `self.timer` is initialized by `bindings::hrtimer_init`.
+///
+#[repr(transparent)]
+#[pin_data(PinnedDrop)]
+pub struct Timer<T> {
+    #[pin]
+    timer: Opaque<bindings::hrtimer>,
+    _t: PhantomData<T>,
+}
+
+// SAFETY: A `Timer` can be moved to other threads and used from there.
+unsafe impl<T> Send for Timer<T> {}
+
+// SAFETY: Timer operatins are locked on C side, so it is safe to operate on a
+// timer from multiple threads
+unsafe impl<T> Sync for Timer<T> {}
+
+impl<T: TimerCallback> Timer<T> {
+    /// Create an initializer for a new `Timer`
+    pub fn new() -> impl PinInit<Self> {
+        crate::pin_init!( Self {
+            timer <- Opaque::ffi_init(move |place: *mut bindings::hrtimer| {
+                // SAFETY: By design of `pin_init!`, `place` is a pointer live
+                // allication. hrtimer_init will initialize `place` and does not
+                // require `place` to be initialized prior to the call.
+                unsafe {
+                    bindings::hrtimer_init(
+                        place,
+                        bindings::CLOCK_MONOTONIC as i32,
+                        bindings::hrtimer_mode_HRTIMER_MODE_REL,
+                    );
+                }
+
+                // SAFETY: `place` is pointing to a live allocation, so the deref
+                // is safe. The `function` field might not be initialized, but
+                // `addr_of_mut` does not create a reference to the field.
+                let function: *mut Option<_> = unsafe { core::ptr::addr_of_mut!((*place).function) };
+
+                // SAFETY: `function` points to a valid allocation.
+                unsafe { core::ptr::write(function, Some(T::Receiver::run)) };
```

Comment:
`place` might contain uninitialized fields, and in particular `function` is not initialized. As far as I know the place projection in the assignee expression `(*place).function` creates a reference. That is UB when the referee is not initialized. `addr_of_mut!()` gets around this. `ptr::write()` is to avoid drop of the destination, which would be UB if it is not initialized. In this particular case, the destination does not have a drop implementation, but we might as well use something that works in the general case.

I tried to find some documentation to back up my claim that the place projection will create a reference internally, but I cannot. The only thing I can find is [the book](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) that states the projection must not violate in-bounds  pointer arithmetic. There is also the [documentation](https://doc.rust-lang.org/std/ptr/macro.addr_of_mut.html) for `addr_of_mut!()` that talks a bit about this, but that is more related to the `&mut expr as *mut _` pattern, which we do not have here. The second example of these docs is similar to this `(*place).field` situation though.

 @y86-dev has [this comment](https://github.com/torvalds/linux/blob/master/rust/kernel/init.rs#L160) in the init module, perhaps they can point to some docs or explain better?

Reply:
Thanks for the explanation, that all makes sense. It does seem weird that it's a three step process to correctly write an uninitialized field - I'll ask opsem if there is a better way, just to see what they say.

Reply:
Ah, here it is - https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#initializing-a-struct-field-by-field. I suppose this could still be simplified to

```rust
addr_of_mut!((*ptr).name).write("Bob".to_string());
```

Seems like this is a more common pattern than I thought, I wonder if it's worth having a macro for this to make the intent more clear

Reply:
the problem with `unsafe { (*ptr).field = value };` is that it runs the destructor of `field`. Since this is a function pointer, there is probably nothing wrong with doing this. but I am not 100% sure. I would just keep it this way.

Reply:
Right, thanks. So `(*ptr).field` is not a problem in itself, but because we want a pointer to `field` we would have to `&mut ((*ptr).field) as *mut _` which would create the reference. `addr_of_mut!()` gets us out of that.